### PR TITLE
Emissive system refactor

### PR DESCRIPTION
**Pull request by TemporalOroboros**
## About The Pull Request

The emissives system is the system that both lets computer screens and vendors glow in the dark and lets mobs and items block those glows. The current implementation relies on using filters to let mobs and items mask out the glow in the dark overlays on some structures. This is difficult to extend properly without massively increasing maptick. This PR changes the emissives system to use BYONDs native image layering to mask emissive overlays. This should prove to be a more extensible option.

tldr; There exists a system that lets computer screens glow on the dark and lets mobs and items block the glow. It isn't very extensible and this PR attempts to make it more extensible.

### How emissive overlays used to work

<details>
<summary>The details (This is a collapsing section. Click me.)</summary>

#### The current system

Currently emissive overlays and the emissive blockers that mask those overlays are handled using a system of inter-masking planes. The emissive overlays and the emissive blockers are placed on separate, hidden plane masters. These are essentially rendering layers and groups. The emissive blocker plane is then used to mask the emissive overlay plane which effectively allows the emissive blockers to block the emissive overlays from being seen. After is has been masked the emissive overlay plane is used to mask the lighting plane, essentially creating holes in the shadows wherever an unblocked glowing thing exists.

#### Why this is a potential problem

This system works fine. In fact it works great! The computer screens glow, any person or item that winds up on a computer blocks the glow, and everything just works. However, this system runs into some issues when you try to extend it to work on things other than structures. Namely, the current system only supports emissive overlays on structures and emissive overlays that are completely unblockable by any means. As a result, several interesting uses to the system require extending the system.

As a result, if you want to apply emissive overlays to items (which exist between structures and mobs) or emissive overlays to turfs (which exist below structures) you must extend the emissives system to get the emissive overlays and emissive blockers to properly function. Doing this naively, by adding extra emissive overlay and emissive blocker planes and applying all of the relevant masking filters, is not exactly performant.

For exactly how unperformant this system is, observe the following graph of maptick.

Credit goes to TheFinalPotato/LemonInTheDark for the graph

Maptick is a major contributor to lag and the higher the maptick the more free lag you, the player, get delivered fresh to your client. See the massive spike? That is the result of #55782, an attempt to add glowing carpet to the game. Since the PR revolved around adding glowing carpet it had to extend the emissives system to allow for emissive turfs and emissive blocking structures. Extending the system was done naively as described above and you can see the results. 1.5 times the maptick across the board. Ouch.

So, we know that extending the system in it's current form is impractical. At least if done naively. Thus we are stuck.

</details>

tldr; The emissive system currently uses inter-plane masking to allow for emissive blockers to function. This is difficult to reasonably extend without murdering maptick. See #56496 for the results of naively extending this system.

### How emissive overlays are going to work

<details>
<summary>The details (This is a collapsing section. Click me.)</summary>

Alright, so we know that the current system of using planes to let the emissive blockers mask the emissive overlays is difficult to extend in it's current form. The solution is to change how the system works so that it can be extended in a more efficient manner. What we want is a system that allows one set of images to be out masked by another set of images and for the first set of images to be capable of masking the light plane. Preferably, we would also like the ability to interleave the masking effect between emissives and emissive blockers with almost arbitrary layering.

Conveniently, this layering and masking is something BYOND already does to normal items and objects. If we put the emissive overlays and the emissive blockers on the same plane we can use their layers to interleave them almost arbitrarily like any normal structures and items! All we need is a way to mask away the emissive blockers from the resulting rendered plane and we can mask the lighting plane with the remaining emissive overlays.

Luckily, BYOND has provided a single filter that is capable of this task. The color matrix filter. This filter can be used to apply a color matrix to an image! Provided that the emissive overlays and the emissive blockers are different colors we can use a color matrix filter to effectively mask out the emissive blockers from the plane! The resulting emissive plane can be applied as an alpha mask to the lighting plane as it used to, to the same effect. The best part is, we get layering practically for free!

This is exactly what this PR does. It converts the emissives system from the old plane and masking based blocking to a new layer-based system which uses BYONDs native layer handling to mask the emissive overlays.

</details>

tldr; This PR changes the emissives system to use BYONDs native image layering so extending it doesn't require additional plane masters.

## Why It's Good For The Game
- Streamlined and significantly more extensible emissive overlays and emissive blockers.
- Possibly enables for the return of neon carpet without murdering maptick.

## Changelog
:cl:
refactor: Emissive overlays and blockers now use BYONDs native image layering.
/:cl:

### CODE REVIEW

File: code/__DEFINES/lighting.dm [hunk 1]
```diff
@@ -75,11 +75,31 @@
...
+#define EM_BLOCK_COLOR list(0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,1, 0,0,0,0)
+/// A globaly cached version of [EM_BLOCK_COLOR] for quick access.
+GLOBAL_LIST_INIT(em_block_color, EM_BLOCK_COLOR)
+/// The color matrix used to mask out emissive blockers on the emissive plane. Alpha should default to zero, be solely dependent on the RGB value of [EMISSIVE_COLOR], and be independant of the RGB value of [EM_BLOCK_COLOR].
+#define EM_MASK_MATRIX list(0,0,0,1/3, 0,0,0,1/3, 0,0,0,1/3, 0,0,0,0, 1,1,1,0)
+/// A globaly cached version of [EM_MASK_MATRIX] for quick access.
+GLOBAL_LIST_INIT(em_mask_matrix, EM_MASK_MATRIX)
+
+/// Produces a mutable appearance glued to the [EMISSIVE_PLANE] dyed to be a mutable appearance.
+/proc/emissive_appearance(icon, icon_state = "", layer = FLOAT_LAYER, alpha = 255, appearance_flags = NONE)
```
Comment (ninjanomnom) [inline code/__DEFINES/lighting.dm:98]: This is not a define

**Review by ninjanomnom — CHANGES_REQUESTED** [review]
Nice work on this, I've got quite a few things I'll want to use this for

File: code/__DEFINES/lighting.dm (hunk 1 above)
Comment (TemporalOroboros) [reply code/__DEFINES/lighting.dm:98]: Helpers then?

**Comment by MrPerson** [comment]
>Provided that the emissive overlays and the emissive blockers are different colors we can use a color matrix filter to effectively mask out the emissive blockers from the plane!

The 100% reliable way to know a great idea like this is that I feel like a moron for not thinking of it first.

**Comment by LemonInTheDark** [comment]
How is this on client performance?

**Comment by TemporalOroboros** [comment]
@LemonInTheDark 
> How is this on client performance?

I'm not sure currently. KylerAce said they were going to profile it when they can get their laptop.

**Comment by LemonInTheDark** [comment]
Guess we could always give it a spin on live and see if anyone complains about client slowdown.
I'm not talking about maptick mind

**Comment by TemporalOroboros** [comment]
Sounds like a plan.

**Comment by LemonInTheDark** [comment]
Any news?

**Comment by TemporalOroboros** [comment]
@LemonInTheDark

KylerAce says that it seems to be fine for maptick and I haven't heard anyone complaining about client performance. It should be good to go provided it doesn't conflict.
